README.md
# bmigraph [![codecov](https://codecov.io/gh/UBC-MDS/bmi-calculator-python/branch/master/graph/badge.svg)](https://codecov.io/gh/UBC-MDS/bmi-calculator-python/tree/master) [![Documentation Status](https://readthedocs.org/projects/bmigraph/badge/?version=latest)](https://bmigraph.readthedocs.io/en/latest/?badge=latest) ![CI/CD Workflow](https://github.com/UBC-MDS/bmi-calculator-python/actions/workflows/ci-cd.yml/badge.svg)

> Authors: Qurat-ul-Ain Azim, Natalie Cho, HanChen Wang, Kelvin Wong
>
> Project complete in accordance with DSCI 524 for the UBC MDS Program 2022-23 for Group 12

This Python package is for calculation of BMI (body mass index), and some more computations based on weight loss goals. The package provides simple answers to a user's weight loss goals in terms of how much weight loss/gain and calorie deficit/increase should be aimed for with a target weight and time frame in mind.  The package also provides helpful visualizations about BMI and calorie intake change trajectory leading to the target.

The latest documentation of the package, hosted on Read the Docs, can be found [here](https://bmigraph.readthedocs.io/en/latest).

## Functions

The package contains the following functions

- `calculate_bmi`: computes user's BMI based on weight and height. Also creates a visual of the BMI on scale
- `project_bmi`: computes how much average change in BMI should be achieved per day given a targeted weight and the timeframe to reach the goal. Also presents a visual trajectory for BMI
- `project_calories`: computes how much average calorie intake is ideal per day given a targeted weight and the timeframe to reach the goal. Also presents a visual trajectory for calories
- `exercise_plan`: suggests possible exercise plans to achieve the targeted weight. Also gives a graph showing how much exercise of each kind is needed per day

## Suitability within Python Ecosystem

Our BMI calculator is unique in the sense that it provides easy and instantly understandable visuals to quickly get the gist of how healthy a person is. The package does not rely on any historical data of a person's weights, and hence needs no dataset files to be provided. The only arguments needed for the functions of this calculator are current weight and height, and target weight with timeframe in case weight change is desired. It also recommends simple figures for weight gain/loss goals. There are many BMI calculators in the Python ecosystem. Some of the examples can be found [here](https://pypi.org/project/body-mass-index/) and [here](https://pypi.org/project/Py-bmi/). Both these offer limited visual aid in understanding one's health metrics and targets related to BMI.


## Installation

```bash
$ pip install bmigraph
```

## Usage

To use the package, import the package with following command:

```
from bmigraph import calculate_bmi, exercise_plan, project_bmi, project_calories
```

To use the functions, see examples below:

### Calculate BMI
```
# Weight 100kg, height 1.85 meters, BMI value returned.
calculate_bmi(100, 1.85, return_graph=False)
# 29.218407596785973

# Weight 100kg, height 1.85 meters, BMI graph returned.
calculate_bmi(100, 1.85, return_graph=True)
```
![BMI graph](_images/calculate_bmi_graph_example.png)

### Compute average BMI change per week
```
# Weight 100kg, height 1.85 meters, BMI goal 25, 30 days to reach goal, return average BMI change per week. 
project_bmi(100, 1.85, 25, 30, return_graph=False)
# -0.98

# Weight 100kg, height 1.85 meters, BMI goal 25, 30 days to reach goal, return plot of Projected BMI trajectory. 
project_bmi(100, 1.85, 25, 30, return_graph=True)
```
![Projected BMI trajectory graph](_images/project_bmi_graph_example.png)

### Compute caloric intake per day based in a target weight
```
# Weight 100kg, height 1.85 meters, male, 25 years old, moderate exercise 3-5 times a week, ideal weight 75kg, 
# 25 days to reach goal, return caloric intake per day based in a target weight. 
project_calories(100, 1.85, 1, 25, 1.6, 75, 25, return_graph=False)
# 2417.0400000000004

# Weight 100kg, height 1.85 meters, male, 25 years old, moderate exercise 3-5 times a week, ideal weight 75kg, 
# 25 days to reach goal, return plot of Projected Weight Loss. 
project_calories(100, 1.85, 1, 25, 1.6, 75, 25, return_graph=True)
```
![Projected Weight Loss graph](_images/project_calories_graph_example.png)

### Create an exercise plan
```
# Weight 100kg, height 1.83 meters, female, aged 27
# Target weight: 68kg in 30 days
exercise_plan(100, 1.83, 2, 27, 68, 30)
# {'Leisure cycling or walking': 213,
# 'Moderate rope-jumping': 88,
# 'General running': 112,
# 'Leisure swimming': 156}

# Weight 100kg, height 1.83 meters, female, aged 27
# Target weight: 68kg in 30 days
# But this time, a graph instead:
exercise_plan(100, 1.83, 2, 27, 68, 30, return_graph=True)
```
![Exercise plan](_images/exercise_plan_graph_example.png)

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a [Code of Conduct](https://github.com/UBC-MDS/bmi-calculator-python/blob/master/CONDUCT.md). You can follow guidelines outlined [here](https://github.com/UBC-MDS/bmi-calculator-python/blob/master/CONTRIBUTING.md) in case you want to contribute to the project. By contributing to this project, you agree to abide by its terms.

The list of contributors to the original project can be found [here](https://github.com/UBC-MDS/bmi-calculator-python/blob/master/CONTRIBUTORS.md).

## License

`bmigraph` was created by Qurat-ul-Ain Azim, Natalie Cho, HanChen Wang, Kelvin Wong. It is licensed under the terms of the MIT license.

## Credits

`bmigraph` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.4.1 (2023-02-04)


## v1.4.0 (2023-01-29)
### Feature
* **ci/cd badge:** Adding a workflow status badge ([`4a423ca`](https://github.com/UBC-MDS/bmi-calculator-python/commit/4a423cab5115c80d802da9001a70aba7f520ef50))

## v1.3.0 (2023-01-28)
### Feature
* Rename package ([`448e2c1`](https://github.com/UBC-MDS/bmi-calculator-python/commit/448e2c1bb0a31c4c3486a6e83d4b682675d1a30f))

## v1.2.1 (2023-01-28)
### Documentation
* Add require.min.js to build ([`7ccacc5`](https://github.com/UBC-MDS/bmi-calculator-python/commit/7ccacc5d3c136ac5de01015b0308a5b12f0409fd))
* Fix image in generated HTML ([`6de1c7b`](https://github.com/UBC-MDS/bmi-calculator-python/commit/6de1c7bf5b64265def83fb4f4fec565177009d01))
* Generate docs/source/*.rst ([`da48710`](https://github.com/UBC-MDS/bmi-calculator-python/commit/da487100ea79dfd661bff2691e87c103d3b4f5a3))

## v1.2.0 (2023-01-25)
### Feature
* **src python scripts:** Added example usage ([`479a4dc`](https://github.com/UBC-MDS/bmi-calculator-python/commit/479a4dc3a41445e542e5d4c0ceac2d5cdc136584))
* **exercise_plan:** Implement ([`d408ab8`](https://github.com/UBC-MDS/bmi-calculator-python/commit/d408ab87365b3f77c408e581b3b7b74a3c1f5395))
* **calculate_bmi:** Add test functions ([`2a055c3`](https://github.com/UBC-MDS/bmi-calculator-python/commit/2a055c361e975c7ab546274944a99855c3b042a6))
* **calculate_bmi:** Add test file. ([`3af057f`](https://github.com/UBC-MDS/bmi-calculator-python/commit/3af057f1ec61ed2cbf70e450432b9aae49610f8a))
* **calculate_bmi:** Updating function implementation with test cases ([`2b64ea9`](https://github.com/UBC-MDS/bmi-calculator-python/commit/2b64ea936af16937544c797bd00bb7ea06b75150))
* **calculate_bmi:** Placeholder ([`661b0e5`](https://github.com/UBC-MDS/bmi-calculator-python/commit/661b0e5b3b2daef4bcf31a6579130614a238ffe6))
* **calculate_bmi:** Add in a placeholder ([`e1b9c98`](https://github.com/UBC-MDS/bmi-calculator-python/commit/e1b9c98f5f64c3d8cca131894e3ee5df276e8c5b))
* **exercise_plan:** Add in a placeholder ([`c0a6e8b`](https://github.com/UBC-MDS/bmi-calculator-python/commit/c0a6e8beb5aed33e1299424304e187f356f03726))

### Fix
* **pyproject.toml:** Deleting pandas and numpy packages. ([`b1d132e`](https://github.com/UBC-MDS/bmi-calculator-python/commit/b1d132e7d4cd0442f9be560c0f7c82114eede757))
* **code and test files:** Fixing codes to make pytest pass. ([`bf9131c`](https://github.com/UBC-MDS/bmi-calculator-python/commit/bf9131cefaf11ae92bdf0ffaf936163c3b5eaf46))
* **calculate_bmi:** Test case ([`33fa22d`](https://github.com/UBC-MDS/bmi-calculator-python/commit/33fa22d245a70f1b157c1f250dc3c2f79ffe8117))

### Documentation
* Update example.ipynb ([`4de6e3d`](https://github.com/UBC-MDS/bmi-calculator-python/commit/4de6e3dc5c6940daab5d86358c8d0b9ea3f35eec))
* Remove references to altair ([`4012618`](https://github.com/UBC-MDS/bmi-calculator-python/commit/4012618333cc0cf456769514f2013400d657f33e))
* Edited contributing.md file to add team member names and add github flow specifications ([`7b07974`](https://github.com/UBC-MDS/bmi-calculator-python/commit/7b079744a878bfff7917c851e9f405a2b8c9ddfe))

## v1.1.0 (2023-01-24)

- Milestone 3 (Prerelease)

## v1.0.0 (2023-01-21)

- Milestone 2

## v0.1.0 (2023-01-14)

- Milestone 1


src/bmi_calculator/exercise_plan.py
import plotly.express as px
from .project_calories import project_calories


def exercise_plan(weight, height, sex, age, target_weight, number_of_days, return_graph=False):
    """Create an exercise plan based on current weight, height, age and target
    BMI.

    The calorie consumption needed each day is computed and a set of activities
    is suggested accordingly. If `return_graph` is `True`, a
    `plotly.graph_objects.Figure` object instead of a dictionary.

    Parameters
    ----------
    weight : float
        Weight, in kilograms.
    height : float
        Height, in meters.
    sex : int
        Sex, used in the Harris–Benedict equation estimation, 1 for male and
        2 for female.
    age : int
        Current age, in years.
    target_weight : float
        Target weight, in kilograms.
    number_of_days : int
        Number of days to reach the target weight.
    return_graph : bool
        Whether to return a graph instead of a dictionary.
    
    Returns
    -------
    dict or plotly.graph_objects.Figure
        If `return_graph` is `False`, we get a dictionary with key to be different
        exercises, and the value to be number of minutes (rounded to nearest
        integer) needed for that activity. Note that each of the activities are
        associated as an "or". If `return_graph` is True, we get a
        `plotly.graph_objects.Figure` instead that can be saved or shared.
    
    Examples
    --------
    >>> exercise_plan(100, 1.83, 2, 27, 68, 30)
    >>> {'Leisure cycling or walking': 213,
    >>> 'Moderate rope-jumping': 88,
    >>> 'General running': 112,
    >>> 'Leisure swimming': 156}
    """
    # Compute daily calorie loss, assume the lowest activity level
    calorie_loss_projection = project_calories(
        weight, height, sex, age, 1.2, target_weight, number_of_days, return_graph=False
    )

    # Use target consumption to calculate the daily activity level
    target_calorie_consumption = 2200 # heuristic
    calorie_consumption_each_day = target_calorie_consumption - calorie_loss_projection
    if calorie_consumption_each_day < 0:
        raise ValueError("Calorie consumption out of range, please verify your numbers")

    # Selected activities simplified and computed from here, it is on per minute:
    # https://www.nutristrategy.com/activitylist4.htm
    activities = {
        "Leisure cycling or walking": 330.0 / 60,
        "Moderate rope-jumping": 800.0 / 60,
        "General running": 630.0 / 60,
        "Leisure swimming": 450.0 / 60
    }

    # Calculate daily activity needed, on an "or" manner
    activity_time = {}
    for _, (activity, calorie_per_min) in enumerate(activities.items()):
        activity_time[activity] = round(calorie_consumption_each_day / calorie_per_min)

    if return_graph:
        graph_bars = {
            "Activity": list(activity_time.keys()),
            "Time, in Minutes": list(activity_time.values())
        }
        return px.bar(graph_bars, x='Time, in Minutes', y='Activity', title="Recommended Daily Activities")
    else:
        return activity_time
